hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java

```diff
@@ -0,0 +1,408 @@
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.IntStream;
+
+import static com.hazelcast.client.HazelcastClient.newHazelcastClient;
+import static com.hazelcast.jet.impl.util.Util.processorToPartitions;
+import static java.util.concurrent.TimeUnit.MICROSECONDS;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * @see Sources#streamMap(String, DistributedPredicate, DistributedFunction, boolean)
+ */
+public final class StreamEventJournalP<E, T> extends AbstractProcessor {
+
+    private static final IdleStrategy IDLER =
+            new BackoffIdleStrategy(0, 0, MICROSECONDS.toNanos(1), MILLISECONDS.toNanos(100));
+    private static final int MIN_FETCH_SIZE = 0;
+    private static final int MAX_FETCH_SIZE = 100;
+
+    private final EventJournalReader<E> eventJournalReader;
+    private final List<Integer> partitions;
+    private final Map<Integer, Long> offsetMap = new HashMap<>();
+    private final SerializablePredicate<E> predicate;
+    private final Projection<E, T> projection;
+    private final boolean startFromLatestSequence;
+
+    private CompletableFuture<Void> jobFuture;
+
+    private StreamEventJournalP(EventJournalReader<E> eventJournalReader,
+                                List<Integer> partitions,
+                                Predicate<E> predicate,
+                                Function<E, T> projectionF,
+                                boolean startFromLatestSequence) {
+        this.eventJournalReader = eventJournalReader;
+        this.partitions = partitions;
+        this.predicate = predicate != null ? predicate::test : null;
+        this.projection = createProjection(projectionF);
+        this.startFromLatestSequence = startFromLatestSequence;
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) throws Exception {
+        jobFuture = context.jobFuture();
+
+        List<ICompletableFuture<EventJournalInitialSubscriberState>> futures = new ArrayList<>();
+        for (Integer partitionId : partitions) {
+            futures.add(eventJournalReader.subscribeToEventJournal(partitionId));
+        }
+        for (int i = 0; i < futures.size(); i++) {
+            ICompletableFuture<EventJournalInitialSubscriberState> future = futures.get(i);
+            offsetMap.put(partitions.get(i), getSequence(future.get()));
+        }
+    }
+
+    private long getSequence(EventJournalInitialSubscriberState state) {
+        if (startFromLatestSequence) {
+            return state.getNewestSequence() == -1 ? 0 : state.getNewestSequence();
+        }
+        return state.getOldestSequence();
+    }
+
+    @Override
+    public boolean complete() {
```

Comment:
regardless of finite(batch) or infinite, a `source` processors should implement just `complete` method. If you never return `true` then it is infinite.